File: frontend/src/lib/api/availability.ts
/**
 * Doctor Availability API Functions
 *
 * NOTE: availability start_time/end_time are naive clinic-local wall-clock
 * times ("HH:MM[:SS]") — see the timezone assumption documented in
 * backend/app/models/doctor_availability.py.
 */

import api from "../api";

export interface AvailabilityWindow {
  id: string;
  doctor_id: string;
  doctor_name?: string | null;
  clinic_id: string | null;
  branch_id: string | null;
  /** 0 = Monday .. 6 = Sunday (matches date.weekday() / JS getDay() - 1 mod 7) */
  weekday: number;
  /** "HH:MM" naive clinic-local time */
  start_time: string;
  /** "HH:MM" naive clinic-local time */
  end_time: string;
  slot_duration_minutes: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateAvailabilityWindowData {
  weekday: number;
  start_time: string; // "HH:MM"
  end_time: string; // "HH:MM"
  slot_duration_minutes?: number;
  clinic_id?: string | null;
  branch_id?: string | null;
  is_active?: boolean;
}

export interface UpdateAvailabilityWindowData {
  weekday?: number;
  start_time?: string;
  end_time?: string;
  slot_duration_minutes?: number;
  clinic_id?: string | null;
  branch_id?: string | null;
  is_active?: boolean;
}

export interface DoctorLeave {
  id: string;
  doctor_id: string;
  /** "YYYY-MM-DD" */
  date: string;
  reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateLeaveData {
  date: string; // "YYYY-MM-DD"
  reason?: string | null;
}

export interface AvailabilitySlot {
  /** ISO datetime of slot start */
  start: string;
  /** ISO datetime of slot end */
  end: string;
  /** "HH:MM" wall-clock start */
  start_time: string;
  /** "HH:MM" wall-clock end */
  end_time: string;
  clinic_id: string | null;
  branch_id: string | null;
}

export interface DoctorSlotsResponse {
  doctor_id: string;
  date: string;
  slots: AvailabilitySlot[];
}

/** List the authenticated doctor's availability windows. */
export async function getMyAvailability(): Promise<AvailabilityWindow[]> {
  const response = await api.get("/api/v1/availability/me");
  return response.data.data;
}

export async function createAvailabilityWindow(
  data: CreateAvailabilityWindowData
): Promise<AvailabilityWindow> {
  const response = await api.post("/api/v1/availability/me", data);
  return response.data;
}

export async function updateAvailabilityWindow(
  id: string,
  data: UpdateAvailabilityWindowData
): Promise<AvailabilityWindow> {
  const response = await api.put(`/api/v1/availability/me/${id}`, data);
  return response.data;
}

export async function deleteAvailabilityWindow(id: string): Promise<void> {
  await api.delete(`/api/v1/availability/me/${id}`);
}

/** List the authenticated doctor's leave days. */
export async function getMyLeaves(): Promise<DoctorLeave[]> {
  const response = await api.get("/api/v1/availability/me/leaves");
  return response.data.data;
}

export async function createLeave(data: CreateLeaveData): Promise<DoctorLeave> {
  const response = await api.post("/api/v1/availability/me/leaves", data);
  return response.data;
}

export async function deleteLeave(id: string): Promise<void> {
  await api.delete(`/api/v1/availability/me/leaves/${id}`);
}

/** List availability windows of all doctors in the active clinic. */
export async function getClinicAvailability(
  clinicId: string
): Promise<AvailabilityWindow[]> {
  const response = await api.get("/api/v1/availability/clinic", {
    headers: { "X-Clinic-Id": clinicId },
  });
  return response.data.data;
}

/** Compute bookable slots for a doctor on a given date (YYYY-MM-DD). */
export async function getDoctorSlots(
  doctorId: string,
  date: string,
  clinicId?: string
): Promise<DoctorSlotsResponse> {
  const params = new URLSearchParams({ date });
  if (clinicId) params.append("clinic_id", clinicId);
  const response = await api.get(
    `/api/v1/availability/doctors/${doctorId}/slots?${params}`
  );
  return response.data;
}

File: backend/app/schemas/availability.py
from datetime import date, datetime, time
from uuid import UUID

from pydantic import BaseModel, Field


class AvailabilityWindowCreate(BaseModel):
    weekday: int = Field(ge=0, le=6)  # 0=Monday .. 6=Sunday (date.weekday())
    start_time: time
    end_time: time
    slot_duration_minutes: int = Field(default=15, ge=5, le=240)
    clinic_id: UUID | None = None
    branch_id: UUID | None = None
    is_active: bool = True


class AvailabilityWindowUpdate(BaseModel):
    weekday: int | None = Field(default=None, ge=0, le=6)
    start_time: time | None = None
    end_time: time | None = None
    slot_duration_minutes: int | None = Field(default=None, ge=5, le=240)
    clinic_id: UUID | None = None
    branch_id: UUID | None = None
    is_active: bool | None = None


class AvailabilityWindowResponse(BaseModel):
    id: UUID
    doctor_id: UUID
    doctor_name: str | None = None
    clinic_id: UUID | None = None
    branch_id: UUID | None = None
    weekday: int
    start_time: time
    end_time: time
    slot_duration_minutes: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class DoctorLeaveCreate(BaseModel):
    date: date
    reason: str | None = Field(default=None, max_length=255)


class DoctorLeaveResponse(BaseModel):
    id: UUID
    doctor_id: UUID
    date: date
    reason: str | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AvailabilitySlot(BaseModel):
    """One bookable slot. start/end are datetimes built from the availability
    window's naive local times (see timezone note in
    models/doctor_availability.py); start_time/end_time are the HH:MM wall
    clock values for display."""

    start: datetime
    end: datetime
    start_time: time
    end_time: time
    clinic_id: UUID | None = None
    branch_id: UUID | None = None


class DoctorSlotsResponse(BaseModel):
    doctor_id: UUID
    date: date
    slots: list[AvailabilitySlot]
